FirebaseFunctionsSwift/Sources/Codable/Callable+Codable.swift

```diff
@@ -0,0 +1,171 @@
+          } else if let error = error {
+            completion(.failure(error))
+          } else {
+            completion(.failure(CallableError.internalError))
+          }
+        } catch {
+          completion(.failure(error))
+        }
+      }
+    } catch {
+      completion(.failure(error))
+    }
+  }
+
+  /// Creates a directly callable function.
+  ///
+  /// This allows users to call a HTTPS Callable Funciton like a normal Swift function:
+  ///
+  ///     let greeter = functions.httpsCallable("greeter",
+  ///                                           requestType: GreetingRequest.self,
+  ///                                           responseType: GreetingResponse.self)
+  ///     try greeter(data) { result in
+  ///       print(result.greeting)
+  ///     }
+  ///
+  /// - Parameters:
+  ///   - data: Parameters to pass to the trigger.
+  ///   - completion: The block to call when the HTTPS request has completed.
+  public func callAsFunction(_ data: Request,
+                             completion: @escaping (Result<Response, Error>)
+                               -> Void) {
+    call(data, completion: completion)
+  }
+
+  #if compiler(>=5.5) && canImport(_Concurrency)
+    /// Executes this Callable HTTPS trigger asynchronously.
+    ///
+    /// The data passed into the trigger must be of the generic `Request` type:
+    ///
+    /// The request to the Cloud Functions backend made by this method automatically includes a
+    /// FCM token to identify the app instance. If a user is logged in with Firebase
+    /// Auth, an auth ID token for the user is also automatically included.
+    ///
+    /// Firebase Cloud Messaging sends data to the Firebase backend periodically to collect information
+    /// regarding the app instance. To stop this, see `Messaging.deleteData()`. It
+    /// resumes with a new FCM Token the next time you call this method.
+    ///
+    /// - Parameter data: The `Request` representing the data to pass to the trigger.
+    ///
+    /// - Throws: An error if any value throws an error during encoding.
+    /// - Throws: An error if any value throws an error during decoding.
+    /// - Throws: An error if the callable fails to complete
+    ///
+    /// - Returns: The decoded `Response` value
+    @available(iOS 15, tvOS 15, macOS 12, watchOS 8, *)
+    public func call(_ data: Request,
+                     encoder: StructureEncoder = StructureEncoder(),
+                     decoder: StructureDecoder =
+                       StructureDecoder()) async throws -> Response {
+      let encoded = try encoder.encode(data)
+      let result = try await callable.call(encoded)
+      return try decoder.decode(Response.self, from: result.data)
+    }
+
+    /// Creates a directly callable function.
+    ///
+    /// This allows users to call a HTTPS Callable Function like a normal Swift function:
+    /// ```swift
+    ///     let greeter = functions.httpsCallable("greeter",
```

Comment:
Now that we support both type-specific parameters and inferred types, shouldn't we add another sample to show how to use this as well?

I.e. add the following:

    /// You can also call a HTTPS Callable function using the following syntax:
    /// ```swift
    ///     let greeter: Callable<GreetingRequest, GreetingResponse> = functions.httpsCallable("greeter")
    ///     let result = try await greeter(data)
    ///     print(result.greeting)
    /// ```

Reply:
Done.